File: backend/app/schemas/background_jobs.py
"""
Background Jobs Schemas

Pydantic schemas for job queue management, task scheduling, retry policies,
progress tracking, and job history.
"""

from typing import Any, Optional
from pydantic import BaseModel, Field
from enum import Enum
from datetime import datetime


# Enums

class JobStatus(str, Enum):
    """Job execution status"""
    PENDING = "pending"
    QUEUED = "queued"
    RUNNING = "running"
    COMPLETED = "completed"
    FAILED = "failed"
    CANCELLED = "cancelled"
    RETRY = "retry"
    TIMEOUT = "timeout"


class JobPriority(str, Enum):
    """Job priority levels"""
    LOW = "low"
    NORMAL = "normal"
    HIGH = "high"
    CRITICAL = "critical"


class JobType(str, Enum):
    """Types of background jobs"""
    QUERY_EXECUTION = "query_execution"
    DATA_EXPORT = "data_export"
    DATA_IMPORT = "data_import"
    REPORT_GENERATION = "report_generation"
    DASHBOARD_REFRESH = "dashboard_refresh"
    CACHE_WARMING = "cache_warming"
    SCHEMA_SYNC = "schema_sync"
    TRANSFORM_EXECUTION = "transform_execution"
    ALERT_CHECK = "alert_check"
    NOTIFICATION_SEND = "notification_send"
    CLEANUP = "cleanup"
    ANALYTICS = "analytics"
    CUSTOM = "custom"


class ScheduleType(str, Enum):
    """Types of job schedules"""
    ONCE = "once"
    INTERVAL = "interval"
    CRON = "cron"
    DAILY = "daily"
    WEEKLY = "weekly"
    MONTHLY = "monthly"


class RetryStrategy(str, Enum):
    """Retry strategies for failed jobs"""
    NONE = "none"
    FIXED = "fixed"
    EXPONENTIAL = "exponential"
    LINEAR = "linear"


class WorkerStatus(str, Enum):
    """Worker status"""
    IDLE = "idle"
    BUSY = "busy"
    PAUSED = "paused"
    OFFLINE = "offline"


# Job Models

class JobCreate(BaseModel):
    """Create a new job"""
    name: str
    job_type: JobType
    priority: JobPriority = JobPriority.NORMAL
    payload: dict[str, Any] = Field(default_factory=dict)
    scheduled_at: Optional[datetime] = None
    timeout_seconds: int = 3600
    retry_policy_id: Optional[str] = None
    workspace_id: Optional[str] = None
    user_id: Optional[str] = None
    tags: list[str] = Field(default_factory=list)
    metadata: dict[str, Any] = Field(default_factory=dict)


class Job(BaseModel):
    """Background job"""
    id: str
    name: str
    job_type: JobType
    status: JobStatus
    priority: JobPriority
    payload: dict[str, Any]
    result: Optional[dict[str, Any]] = None
    error_message: Optional[str] = None
    error_details: Optional[dict[str, Any]] = None
    progress: float = 0.0  # 0-100
    progress_message: Optional[str] = None
    scheduled_at: Optional[datetime] = None
    started_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None
    timeout_seconds: int
    retry_count: int = 0
    max_retries: int = 0
    retry_policy_id: Optional[str] = None
    worker_id: Optional[str] = None
    workspace_id: Optional[str] = None
    user_id: Optional[str] = None
    tags: list[str] = Field(default_factory=list)
    metadata: dict[str, Any] = Field(default_factory=dict)
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)


class JobUpdate(BaseModel):
    """Update job fields"""
    status: Optional[JobStatus] = None
    priority: Optional[JobPriority] = None
    progress: Optional[float] = None
    progress_message: Optional[str] = None
    result: Optional[dict[str, Any]] = None
    error_message: Optional[str] = None
    tags: Optional[list[str]] = None


class JobProgress(BaseModel):
    """Job progress update"""
    job_id: str
    progress: float
    message: Optional[str] = None
    metadata: Optional[dict[str, Any]] = None


class JobResult(BaseModel):
    """Job execution result"""
    job_id: str
    status: JobStatus
    result: Optional[dict[str, Any]] = None
    error_message: Optional[str] = None
    error_details: Optional[dict[str, Any]] = None
    execution_time_ms: float
    completed_at: datetime


# Schedule Models

class ScheduleCreate(BaseModel):
    """Create a scheduled job"""
    name: str
    description: Optional[str] = None
    job_type: JobType
    schedule_type: ScheduleType
    schedule_config: dict[str, Any]  # cron expression, interval, etc.
    payload: dict[str, Any] = Field(default_factory=dict)
    priority: JobPriority = JobPriority.NORMAL
    timeout_seconds: int = 3600
    retry_policy_id: Optional[str] = None
    enabled: bool = True
    workspace_id: Optional[str] = None
    tags: list[str] = Field(default_factory=list)


class Schedule(BaseModel):
    """Scheduled job definition"""
    id: str
    name: str
    description: Optional[str] = None
    job_type: JobType
    schedule_type: ScheduleType
    schedule_config: dict[str, Any]
    payload: dict[str, Any]
    priority: JobPriority
    timeout_seconds: int
    retry_policy_id: Optional[str] = None
    enabled: bool
    last_run_at: Optional[datetime] = None
    next_run_at: Optional[datetime] = None
    last_run_status: Optional[JobStatus] = None
    run_count: int = 0
    failure_count: int = 0
    workspace_id: Optional[str] = None
    tags: list[str] = Field(default_factory=list)
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)


class ScheduleUpdate(BaseModel):
    """Update schedule fields"""
    name: Optional[str] = None
    description: Optional[str] = None
    schedule_type: Optional[ScheduleType] = None
    schedule_config: Optional[dict[str, Any]] = None
    payload: Optional[dict[str, Any]] = None
    priority: Optional[JobPriority] = None
    timeout_seconds: Optional[int] = None
    retry_policy_id: Optional[str] = None
    enabled: Optional[bool] = None
    tags: Optional[list[str]] = None


class ScheduleRun(BaseModel):
    """Record of a scheduled job run"""
    id: str
    schedule_id: str
    job_id: str
    status: JobStatus
    scheduled_time: datetime
    started_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None
    execution_time_ms: Optional[float] = None
    error_message: Optional[str] = None


# Retry Policy Models

class RetryPolicyCreate(BaseModel):
    """Create retry policy"""
    name: str
    description: Optional[str] = None
    strategy: RetryStrategy
    max_retries: int = 3
    initial_delay_seconds: int = 60
    max_delay_seconds: int = 3600
    multiplier: float = 2.0  # For exponential backoff
    retry_on_statuses: list[str] = Field(default_factory=lambda: ["failed", "timeout"])
    retry_on_errors: list[str] = Field(default_factory=list)  # Error code patterns


class RetryPolicy(BaseModel):
    """Retry policy definition"""
    id: str
    name: str
    description: Optional[str] = None
    strategy: RetryStrategy
    max_retries: int
    initial_delay_seconds: int
    max_delay_seconds: int
    multiplier: float
    retry_on_statuses: list[str]
    retry_on_errors: list[str]
    created_at: datetime = Field(default_factory=datetime.utcnow)


class RetryPolicyUpdate(BaseModel):
    """Update retry policy"""
    name: Optional[str] = None
    description: Optional[str] = None
    strategy: Optional[RetryStrategy] = None
    max_retries: Optional[int] = None
    initial_delay_seconds: Optional[int] = None
    max_delay_seconds: Optional[int] = None
    multiplier: Optional[float] = None
    retry_on_statuses: Optional[list[str]] = None
    retry_on_errors: Optional[list[str]] = None


# Worker Models

class Worker(BaseModel):
    """Background job worker"""
    id: str
    name: str
    hostname: str
    status: WorkerStatus
    current_job_id: Optional[str] = None
    job_types: list[JobType]
    max_concurrent_jobs: int = 1
    jobs_processed: int = 0
    jobs_failed: int = 0
    last_heartbeat: datetime
    started_at: datetime
    metadata: dict[str, Any] = Field(default_factory=dict)


class WorkerHeartbeat(BaseModel):
    """Worker heartbeat update"""
    worker_id: str
    status: WorkerStatus
    current_job_id: Optional[str] = None
    memory_usage_mb: Optional[float] = None
    cpu_usage_percent: Optional[float] = None


# Queue Models

class QueueStats(BaseModel):
    """Job queue statistics"""
    name: str
    pending_count: int
    running_count: int
    completed_count: int
    failed_count: int
    avg_wait_time_ms: float
    avg_execution_time_ms: float
    throughput_per_minute: float
    oldest_pending_age_seconds: float


class QueueConfig(BaseModel):
    """Queue configuration"""
    name: str
    job_types: list[JobType]
    max_workers: int = 4
    max_queue_size: int = 10000
    default_timeout_seconds: int = 3600
    default_priority: JobPriority = JobPriority.NORMAL
    rate_limit_per_minute: Optional[int] = None


# Job History Models

class JobHistoryEntry(BaseModel):
    """Job history entry"""
    id: str
    job_id: str
    job_name: str
    job_type: JobType
    status: JobStatus
    priority: JobPriority
    workspace_id: Optional[str] = None
    user_id: Optional[str] = None
    execution_time_ms: Optional[float] = None
    retry_count: int = 0
    error_message: Optional[str] = None
    scheduled_at: Optional[datetime] = None
    started_at: Optional[datetime] = None
    completed_at: datetime
    tags: list[str] = Field(default_factory=list)


class JobHistoryStats(BaseModel):
    """Job history statistics"""
    total_jobs: int
    completed_jobs: int
    failed_jobs: int
    cancelled_jobs: int
    avg_execution_time_ms: float
    total_execution_time_ms: float
    success_rate: float
    by_type: dict[str, int]
    by_status: dict[str, int]
    by_hour: list[dict[str, Any]]
    slowest_jobs: list[JobHistoryEntry]


# Response Models

class JobListResponse(BaseModel):
    """List jobs response"""
    jobs: list[Job]
    total: int


class ScheduleListResponse(BaseModel):
    """List schedules response"""
    schedules: list[Schedule]
    total: int


class RetryPolicyListResponse(BaseModel):
    """List retry policies response"""
    policies: list[RetryPolicy]
    total: int


class WorkerListResponse(BaseModel):
    """List workers response"""
    workers: list[Worker]
    total: int


class JobHistoryListResponse(BaseModel):
    """List job history response"""
    entries: list[JobHistoryEntry]
    total: int


class ScheduleRunListResponse(BaseModel):
    """List schedule runs response"""
    runs: list[ScheduleRun]
    total: int


# Dashboard Models

class JobsDashboard(BaseModel):
    """Jobs dashboard summary"""
    active_jobs: int
    pending_jobs: int
    completed_today: int
    failed_today: int
    success_rate: float
    avg_wait_time_ms: float
    avg_execution_time_ms: float
    active_workers: int
    total_workers: int
    queues: list[QueueStats]
    recent_failures: list[Job]


# Configuration

class JobsConfig(BaseModel):
    """Background jobs configuration"""
    enabled: bool = True
    max_workers: int = 4
    default_timeout_seconds: int = 3600
    max_retry_attempts: int = 3
    cleanup_completed_after_hours: int = 168  # 1 week
    cleanup_failed_after_hours: int = 720  # 30 days
    rate_limit_enabled: bool = True
    rate_limit_per_minute: int = 100


class JobsConfigUpdate(BaseModel):
    """Update jobs configuration"""
    enabled: Optional[bool] = None
    max_workers: Optional[int] = None
    default_timeout_seconds: Optional[int] = None
    max_retry_attempts: Optional[int] = None
    cleanup_completed_after_hours: Optional[int] = None
    cleanup_failed_after_hours: Optional[int] = None
    rate_limit_enabled: Optional[bool] = None
    rate_limit_per_minute: Optional[int] = None


# Constants

JOB_TYPE_LABELS = {
    JobType.QUERY_EXECUTION: "Query Execution",
    JobType.DATA_EXPORT: "Data Export",
    JobType.DATA_IMPORT: "Data Import",
    JobType.REPORT_GENERATION: "Report Generation",
    JobType.DASHBOARD_REFRESH: "Dashboard Refresh",
    JobType.CACHE_WARMING: "Cache Warming",
    JobType.SCHEMA_SYNC: "Schema Sync",
    JobType.TRANSFORM_EXECUTION: "Transform Execution",
    JobType.ALERT_CHECK: "Alert Check",
    JobType.NOTIFICATION_SEND: "Notification Send",
    JobType.CLEANUP: "Cleanup",
    JobType.ANALYTICS: "Analytics",
    JobType.CUSTOM: "Custom",
}

JOB_STATUS_LABELS = {
    JobStatus.PENDING: "Pending",
    JobStatus.QUEUED: "Queued",
    JobStatus.RUNNING: "Running",
    JobStatus.COMPLETED: "Completed",
    JobStatus.FAILED: "Failed",
    JobStatus.CANCELLED: "Cancelled",
    JobStatus.RETRY: "Retry",
    JobStatus.TIMEOUT: "Timeout",
}

PRIORITY_ORDER = {
    JobPriority.CRITICAL: 0,
    JobPriority.HIGH: 1,
    JobPriority.NORMAL: 2,
    JobPriority.LOW: 3,
}


# Helper Functions

def calculate_next_retry_delay(
    retry_count: int,
    policy: RetryPolicy,
) -> int:
    """Calculate delay before next retry."""
    if policy.strategy == RetryStrategy.NONE:
        return 0
    elif policy.strategy == RetryStrategy.FIXED:
        delay = policy.initial_delay_seconds
    elif policy.strategy == RetryStrategy.LINEAR:
        delay = policy.initial_delay_seconds * (retry_count + 1)
    else:  # exponential
        delay = policy.initial_delay_seconds * (policy.multiplier ** retry_count)

    return min(int(delay), policy.max_delay_seconds)


def parse_cron_expression(cron: str) -> dict[str, Any]:
    """Parse cron expression into components."""
    parts = cron.split()
    if len(parts) != 5:
        raise ValueError("Invalid cron expression: expected 5 parts")

    return {
        "minute": parts[0],
        "hour": parts[1],
        "day_of_month": parts[2],
        "month": parts[3],
        "day_of_week": parts[4],
    }


def get_priority_order(priority: JobPriority) -> int:
    """Get numeric order for priority sorting."""
    return PRIORITY_ORDER.get(priority, 2)


def is_terminal_status(status: JobStatus) -> bool:
    """Check if job status is terminal (won't change)."""
    return status in [
        JobStatus.COMPLETED,
        JobStatus.FAILED,
        JobStatus.CANCELLED,
    ]


def can_retry(job: Job, policy: Optional[RetryPolicy]) -> bool:
    """Check if job can be retried."""
    if not policy or policy.strategy == RetryStrategy.NONE:
        return False
    if job.retry_count >= policy.max_retries:
        return False
    if job.status.value not in policy.retry_on_statuses:
        return False
    return True

File: frontend/src/types/background_jobs.ts
/**
 * Background Jobs Types
 *
 * TypeScript types for job queue management, task scheduling, retry policies,
 * progress tracking, and job history.
 */

// Enums

export enum JobStatus {
  PENDING = 'pending',
  QUEUED = 'queued',
  RUNNING = 'running',
  COMPLETED = 'completed',
  FAILED = 'failed',
  CANCELLED = 'cancelled',
  RETRY = 'retry',
  TIMEOUT = 'timeout',
}

export enum JobPriority {
  LOW = 'low',
  NORMAL = 'normal',
  HIGH = 'high',
  CRITICAL = 'critical',
}

export enum JobType {
  QUERY_EXECUTION = 'query_execution',
  DATA_EXPORT = 'data_export',
  DATA_IMPORT = 'data_import',
  REPORT_GENERATION = 'report_generation',
  DASHBOARD_REFRESH = 'dashboard_refresh',
  CACHE_WARMING = 'cache_warming',
  SCHEMA_SYNC = 'schema_sync',
  TRANSFORM_EXECUTION = 'transform_execution',
  ALERT_CHECK = 'alert_check',
  NOTIFICATION_SEND = 'notification_send',
  CLEANUP = 'cleanup',
  ANALYTICS = 'analytics',
  CUSTOM = 'custom',
}

export enum ScheduleType {
  ONCE = 'once',
  INTERVAL = 'interval',
  CRON = 'cron',
  DAILY = 'daily',
  WEEKLY = 'weekly',
  MONTHLY = 'monthly',
}

export enum RetryStrategy {
  NONE = 'none',
  FIXED = 'fixed',
  EXPONENTIAL = 'exponential',
  LINEAR = 'linear',
}

export enum WorkerStatus {
  IDLE = 'idle',
  BUSY = 'busy',
  PAUSED = 'paused',
  OFFLINE = 'offline',
}

// Job Types

export interface Job {
  id: string;
  name: string;
  job_type: JobType;
  status: JobStatus;
  priority: JobPriority;
  payload: Record<string, unknown>;
  result?: Record<string, unknown> | null;
  error_message?: string | null;
  error_details?: Record<string, unknown> | null;
  progress: number;
  progress_message?: string | null;
  scheduled_at?: string | null;
  started_at?: string | null;
  completed_at?: string | null;
  timeout_seconds: number;
  retry_count: number;
  max_retries: number;
  retry_policy_id?: string | null;
  worker_id?: string | null;
  workspace_id?: string | null;
  user_id?: string | null;
  tags: string[];
  metadata: Record<string, unknown>;
  created_at: string;
  updated_at: string;
}

export interface JobCreate {
  name: string;
  job_type: JobType;
  priority?: JobPriority;
  payload?: Record<string, unknown>;
  scheduled_at?: string;
  timeout_seconds?: number;
  retry_policy_id?: string;
  workspace_id?: string;
  user_id?: string;
  tags?: string[];
  metadata?: Record<string, unknown>;
}

export interface JobUpdate {
  status?: JobStatus;
  priority?: JobPriority;
  progress?: number;
  progress_message?: string;
  result?: Record<string, unknown>;
  error_message?: string;
  tags?: string[];
}

export interface JobProgress {
  job_id: string;
  progress: number;
  message?: string;
  metadata?: Record<string, unknown>;
}

export interface JobResult {
  job_id: string;
  status: JobStatus;
  result?: Record<string, unknown>;
  error_message?: string;
  error_details?: Record<string, unknown>;
  execution_time_ms: number;
  completed_at: string;
}

// Schedule Types

export interface Schedule {
  id: string;
  name: string;
  description?: string | null;
  job_type: JobType;
  schedule_type: ScheduleType;
  schedule_config: Record<string, unknown>;
  payload: Record<string, unknown>;
  priority: JobPriority;
  timeout_seconds: number;
  retry_policy_id?: string | null;
  enabled: boolean;
  last_run_at?: string | null;
  next_run_at?: string | null;
  last_run_status?: JobStatus | null;
  run_count: number;
  failure_count: number;
  workspace_id?: string | null;
  tags: string[];
  created_at: string;
  updated_at: string;
}

export interface ScheduleCreate {
  name: string;
  description?: string;
  job_type: JobType;
  schedule_type: ScheduleType;
  schedule_config: Record<string, unknown>;
  payload?: Record<string, unknown>;
  priority?: JobPriority;
  timeout_seconds?: number;
  retry_policy_id?: string;
  enabled?: boolean;
  workspace_id?: string;
  tags?: string[];
}

export interface ScheduleUpdate {
  name?: string;
  description?: string;
  schedule_type?: ScheduleType;
  schedule_config?: Record<string, unknown>;
  payload?: Record<string, unknown>;
  priority?: JobPriority;
  timeout_seconds?: number;
  retry_policy_id?: string;
  enabled?: boolean;
  tags?: string[];
}

export interface ScheduleRun {
  id: string;
  schedule_id: string;
  job_id: string;
  status: JobStatus;
  scheduled_time: string;
  started_at?: string | null;
  completed_at?: string | null;
  execution_time_ms?: number | null;
  error_message?: string | null;
}

// Retry Policy Types

export interface RetryPolicy {
  id: string;
  name: string;
  description?: string | null;
  strategy: RetryStrategy;
  max_retries: number;
  initial_delay_seconds: number;
  max_delay_seconds: number;
  multiplier: number;
  retry_on_statuses: string[];
  retry_on_errors: string[];
  created_at: string;
}

export interface RetryPolicyCreate {
  name: string;
  description?: string;
  strategy: RetryStrategy;
  max_retries?: number;
  initial_delay_seconds?: number;
  max_delay_seconds?: number;
  multiplier?: number;
  retry_on_statuses?: string[];
  retry_on_errors?: string[];
}

export interface RetryPolicyUpdate {
  name?: string;
  description?: string;
  strategy?: RetryStrategy;
  max_retries?: number;
  initial_delay_seconds?: number;
  max_delay_seconds?: number;
  multiplier?: number;
  retry_on_statuses?: string[];
  retry_on_errors?: string[];
}

// Worker Types

export interface Worker {
  id: string;
  name: string;
  hostname: string;
  status: WorkerStatus;
  current_job_id?: string | null;
  job_types: JobType[];
  max_concurrent_jobs: number;
  jobs_processed: number;
  jobs_failed: number;
  last_heartbeat: string;
  started_at: string;
  metadata: Record<string, unknown>;
}

export interface WorkerHeartbeat {
  worker_id: string;
  status: WorkerStatus;
  current_job_id?: string;
  memory_usage_mb?: number;
  cpu_usage_percent?: number;
}

// Queue Types

export interface QueueStats {
  name: string;
  pending_count: number;
  running_count: number;
  completed_count: number;
  failed_count: number;
  avg_wait_time_ms: number;
  avg_execution_time_ms: number;
  throughput_per_minute: number;
  oldest_pending_age_seconds: number;
}

export interface QueueConfig {
  name: string;
  job_types: JobType[];
  max_workers: number;
  max_queue_size: number;
  default_timeout_seconds: number;
  default_priority: JobPriority;
  rate_limit_per_minute?: number;
}

// History Types

export interface JobHistoryEntry {
  id: string;
  job_id: string;
  job_name: string;
  job_type: JobType;
  status: JobStatus;
  priority: JobPriority;
  workspace_id?: string | null;
  user_id?: string | null;
  execution_time_ms?: number | null;
  retry_count: number;
  error_message?: string | null;
  scheduled_at?: string | null;
  started_at?: string | null;
  completed_at: string;
  tags: string[];
}

export interface JobHistoryStats {
  total_jobs: number;
  completed_jobs: number;
  failed_jobs: number;
  cancelled_jobs: number;
  avg_execution_time_ms: number;
  total_execution_time_ms: number;
  success_rate: number;
  by_type: Record<string, number>;
  by_status: Record<string, number>;
  by_hour: Array<Record<string, unknown>>;
  slowest_jobs: JobHistoryEntry[];
}

// Dashboard Types

export interface JobsDashboard {
  active_jobs: number;
  pending_jobs: number;
  completed_today: number;
  failed_today: number;
  success_rate: number;
  avg_wait_time_ms: number;
  avg_execution_time_ms: number;
  active_workers: number;
  total_workers: number;
  queues: QueueStats[];
  recent_failures: Job[];
}

// Configuration Types

export interface JobsConfig {
  enabled: boolean;
  max_workers: number;
  default_timeout_seconds: number;
  max_retry_attempts: number;
  cleanup_completed_after_hours: number;
  cleanup_failed_after_hours: number;
  rate_limit_enabled: boolean;
  rate_limit_per_minute: number;
}

export interface JobsConfigUpdate {
  enabled?: boolean;
  max_workers?: number;
  default_timeout_seconds?: number;
  max_retry_attempts?: number;
  cleanup_completed_after_hours?: number;
  cleanup_failed_after_hours?: number;
  rate_limit_enabled?: boolean;
  rate_limit_per_minute?: number;
}

// Response Types

export interface JobListResponse {
  jobs: Job[];
  total: number;
}

export interface ScheduleListResponse {
  schedules: Schedule[];
  total: number;
}

export interface RetryPolicyListResponse {
  policies: RetryPolicy[];
  total: number;
}

export interface WorkerListResponse {
  workers: Worker[];
  total: number;
}

export interface JobHistoryListResponse {
  entries: JobHistoryEntry[];
  total: number;
}

export interface ScheduleRunListResponse {
  runs: ScheduleRun[];
  total: number;
}

// Constants

export const JOB_TYPE_LABELS: Record<JobType, string> = {
  [JobType.QUERY_EXECUTION]: 'Query Execution',
  [JobType.DATA_EXPORT]: 'Data Export',
  [JobType.DATA_IMPORT]: 'Data Import',
  [JobType.REPORT_GENERATION]: 'Report Generation',
  [JobType.DASHBOARD_REFRESH]: 'Dashboard Refresh',
  [JobType.CACHE_WARMING]: 'Cache Warming',
  [JobType.SCHEMA_SYNC]: 'Schema Sync',
  [JobType.TRANSFORM_EXECUTION]: 'Transform Execution',
  [JobType.ALERT_CHECK]: 'Alert Check',
  [JobType.NOTIFICATION_SEND]: 'Notification Send',
  [JobType.CLEANUP]: 'Cleanup',
  [JobType.ANALYTICS]: 'Analytics',
  [JobType.CUSTOM]: 'Custom',
};

export const JOB_STATUS_LABELS: Record<JobStatus, string> = {
  [JobStatus.PENDING]: 'Pending',
  [JobStatus.QUEUED]: 'Queued',
  [JobStatus.RUNNING]: 'Running',
  [JobStatus.COMPLETED]: 'Completed',
  [JobStatus.FAILED]: 'Failed',
  [JobStatus.CANCELLED]: 'Cancelled',
  [JobStatus.RETRY]: 'Retry',
  [JobStatus.TIMEOUT]: 'Timeout',
};

export const JOB_STATUS_COLORS: Record<JobStatus, string> = {
  [JobStatus.PENDING]: 'gray',
  [JobStatus.QUEUED]: 'blue',
  [JobStatus.RUNNING]: 'indigo',
  [JobStatus.COMPLETED]: 'green',
  [JobStatus.FAILED]: 'red',
  [JobStatus.CANCELLED]: 'gray',
  [JobStatus.RETRY]: 'yellow',
  [JobStatus.TIMEOUT]: 'orange',
};

export const JOB_PRIORITY_LABELS: Record<JobPriority, string> = {
  [JobPriority.LOW]: 'Low',
  [JobPriority.NORMAL]: 'Normal',
  [JobPriority.HIGH]: 'High',
  [JobPriority.CRITICAL]: 'Critical',
};

export const JOB_PRIORITY_COLORS: Record<JobPriority, string> = {
  [JobPriority.LOW]: 'gray',
  [JobPriority.NORMAL]: 'blue',
  [JobPriority.HIGH]: 'yellow',
  [JobPriority.CRITICAL]: 'red',
};

export const SCHEDULE_TYPE_LABELS: Record<ScheduleType, string> = {
  [ScheduleType.ONCE]: 'Once',
  [ScheduleType.INTERVAL]: 'Interval',
  [ScheduleType.CRON]: 'Cron',
  [ScheduleType.DAILY]: 'Daily',
  [ScheduleType.WEEKLY]: 'Weekly',
  [ScheduleType.MONTHLY]: 'Monthly',
};

export const RETRY_STRATEGY_LABELS: Record<RetryStrategy, string> = {
  [RetryStrategy.NONE]: 'None',
  [RetryStrategy.FIXED]: 'Fixed Delay',
  [RetryStrategy.EXPONENTIAL]: 'Exponential Backoff',
  [RetryStrategy.LINEAR]: 'Linear Backoff',
};

export const WORKER_STATUS_LABELS: Record<WorkerStatus, string> = {
  [WorkerStatus.IDLE]: 'Idle',
  [WorkerStatus.BUSY]: 'Busy',
  [WorkerStatus.PAUSED]: 'Paused',
  [WorkerStatus.OFFLINE]: 'Offline',
};

export const WORKER_STATUS_COLORS: Record<WorkerStatus, string> = {
  [WorkerStatus.IDLE]: 'green',
  [WorkerStatus.BUSY]: 'blue',
  [WorkerStatus.PAUSED]: 'yellow',
  [WorkerStatus.OFFLINE]: 'gray',
};

// Helper Functions

export function formatDuration(ms: number): string {
  if (ms < 1000) return `${ms.toFixed(0)}ms`;
  if (ms < 60000) return `${(ms / 1000).toFixed(1)}s`;
  if (ms < 3600000) return `${(ms / 60000).toFixed(1)}m`;
  return `${(ms / 3600000).toFixed(1)}h`;
}

export function formatSchedule(schedule: Schedule): string {
  const { schedule_type, schedule_config } = schedule;

  switch (schedule_type) {
    case ScheduleType.ONCE:
      return `Once at ${schedule_config.run_at}`;
    case ScheduleType.INTERVAL:
      const seconds = schedule_config.interval_seconds as number;
      if (seconds < 60) return `Every ${seconds}s`;
      if (seconds < 3600) return `Every ${seconds / 60}m`;
      return `Every ${seconds / 3600}h`;
    case ScheduleType.DAILY:
      return `Daily at ${String(schedule_config.hour).padStart(2, '0')}:${String(schedule_config.minute || 0).padStart(2, '0')}`;
    case ScheduleType.WEEKLY:
      const days = ['Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat', 'Sun'];
      return `Weekly on ${days[schedule_config.day_of_week as number]}`;
    case ScheduleType.MONTHLY:
      return `Monthly on day ${schedule_config.day}`;
    case ScheduleType.CRON:
      return `Cron: ${schedule_config.expression}`;
    default:
      return 'Unknown';
  }
}

export function getJobStatusColor(status: JobStatus): string {
  return JOB_STATUS_COLORS[status] || 'gray';
}

export function getJobPriorityColor(priority: JobPriority): string {
  return JOB_PRIORITY_COLORS[priority] || 'gray';
}

export function getWorkerStatusColor(status: WorkerStatus): string {
  return WORKER_STATUS_COLORS[status] || 'gray';
}

export function isJobRunning(job: Job): boolean {
  return job.status === JobStatus.RUNNING;
}

export function isJobComplete(job: Job): boolean {
  return [JobStatus.COMPLETED, JobStatus.FAILED, JobStatus.CANCELLED].includes(job.status);
}

export function canRetryJob(job: Job): boolean {
  return [JobStatus.FAILED, JobStatus.TIMEOUT].includes(job.status);
}

export function canCancelJob(job: Job): boolean {
  return [JobStatus.PENDING, JobStatus.QUEUED, JobStatus.RUNNING].includes(job.status);
}

export function calculateRetryDelay(
  retryCount: number,
  strategy: RetryStrategy,
  initialDelay: number,
  multiplier: number,
  maxDelay: number
): number {
  let delay: number;

  switch (strategy) {
    case RetryStrategy.NONE:
      return 0;
    case RetryStrategy.FIXED:
      delay = initialDelay;
      break;
    case RetryStrategy.LINEAR:
      delay = initialDelay * (retryCount + 1);
      break;
    case RetryStrategy.EXPONENTIAL:
      delay = initialDelay * Math.pow(multiplier, retryCount);
      break;
    default:
      delay = initialDelay;
  }

  return Math.min(delay, maxDelay);
}

// State Management

export interface BackgroundJobsState {
  jobs: Job[];
  schedules: Schedule[];
  retryPolicies: RetryPolicy[];
  workers: Worker[];
  history: JobHistoryEntry[];
  historyStats: JobHistoryStats | null;
  dashboard: JobsDashboard | null;
  config: JobsConfig | null;
  selectedJob: Job | null;
  selectedSchedule: Schedule | null;
  isLoading: boolean;
  error: string | null;
}

export function createInitialBackgroundJobsState(): BackgroundJobsState {
  return {
    jobs: [],
    schedules: [],
    retryPolicies: [],
    workers: [],
    history: [],
    historyStats: null,
    dashboard: null,
    config: null,
    selectedJob: null,
    selectedSchedule: null,
    isLoading: false,
    error: null,
  };
}
